pkg/controllers/user/monitoring/config_refresh_handler.go

```diff
@@ -0,0 +1,137 @@
+package monitoring
+
+import (
+	"encoding/json"
+	"reflect"
+	"sort"
+
+	"github.com/rancher/rancher/pkg/controllers/user/nslabels"
+	apiv1 "github.com/rancher/types/apis/core/v1"
+	rmonitoringv1 "github.com/rancher/types/apis/monitoring.coreos.com/v1"
+
+	monitoringv1 "github.com/coreos/prometheus-operator/pkg/client/monitoring/v1"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/tools/cache"
+)
+
+const (
+	projectNSAnnotation        = "project.cattle.io/namespaces"
+	nsByProjectIndex           = "monitoring.cluster.cattle.io/ns-by-project"
+	promByMemberNamespaceIndex = "monitoring.cluster.cattle.io/prom-by-member-ns"
+)
+
+type ConfigRefreshHandler struct {
+	prometheusClient  rmonitoringv1.PrometheusInterface
+	prometheusIndexer cache.Indexer
+	nsIndexer         cache.Indexer
+}
+
+func (h *ConfigRefreshHandler) syncNamespace(key string, obj *corev1.Namespace) (runtime.Object, error) {
+	if obj == nil || obj.DeletionTimestamp != nil {
+		return obj, nil
+	}
+	promList, err := h.prometheusIndexer.ByIndex(promByMemberNamespaceIndex, obj.Name)
+	if err != nil {
+		return obj, err
+	}
+
+	for _, object := range promList {
+		prometheus, ok := object.(*monitoringv1.Prometheus)
+		if !ok {
+			continue
+		}
+		h.prometheusClient.Controller().Enqueue(prometheus.Namespace, prometheus.Name)
+	}
+
+	return obj, nil
+}
+
+func (h *ConfigRefreshHandler) syncPrometheus(key string, obj *monitoringv1.Prometheus) (runtime.Object, error) {
+	if obj == nil || obj.DeletionTimestamp != nil {
+		return obj, nil
+	}
+
+	nsObject, exists, err := h.nsIndexer.GetByKey(obj.Namespace)
```

Comment:
for just getting by name, couldnt you just use a normal lister? should only use indexer when using a custom index